Mathlib/Data/Matroid/Rank/ENat.lean

```diff
@@ -298,6 +261,170 @@ lemma eRk_inter_add_eRk_union_le (M : Matroid α) (X Y : Set α) :
 
 alias eRk_submod := eRk_inter_add_eRk_union_le
 
+lemma eRk_submod_insert (M : Matroid α) (X Y : Set α) :
+    M.eRk (insert e (X ∩ Y)) + M.eRk (insert e (X ∪ Y))
+      ≤ M.eRk (insert e X) + M.eRk (insert e Y) := by
+  rw [insert_inter_distrib, insert_union_distrib]
+  apply M.eRk_submod
+
+lemma eRk_submod_compl (M : Matroid α) (X Y : Set α) :
+    M.eRk (M.E \ (X ∪ Y)) + M.eRk (M.E \ (X ∩ Y)) ≤ M.eRk (M.E \ X) + M.eRk (M.E \ Y) := by
+  rw [← diff_inter_diff, diff_inter]
+  apply M.eRk_submod
+
+lemma eRk_submod_insert_compl (M : Matroid α) (X Y : Set α) :
+    M.eRk (M.E \ insert e (X ∪ Y)) + M.eRk (M.E \ insert e (X ∩ Y)) ≤
+      M.eRk (M.E \ insert e X) + M.eRk (M.E \ insert e Y) := by
+  rw [insert_union_distrib, insert_inter_distrib]
+  exact M.eRk_submod_compl (insert e X) (insert e Y)
+
+lemma eRk_union_le_eRk_add_eRk (M : Matroid α) (X Y : Set α) : M.eRk (X ∪ Y) ≤ M.eRk X + M.eRk Y :=
+  le_add_self.trans (M.eRk_submod X Y)
+
+lemma eRk_eq_eRk_union_eRk_le_zero (X : Set α) (hY : M.eRk Y ≤ 0) : M.eRk (X ∪ Y) = M.eRk X :=
+  (((M.eRk_union_le_eRk_add_eRk X Y).trans (add_le_add_left hY _)).trans_eq (add_zero _)).antisymm
+    (M.eRk_mono subset_union_left)
+
+lemma eRk_eq_eRk_diff_eRk_le_zero (X : Set α) (hY : M.eRk Y ≤ 0) : M.eRk (X \ Y) = M.eRk X := by
+  rw [← eRk_eq_eRk_union_eRk_le_zero (X \ Y) hY, diff_union_self, eRk_eq_eRk_union_eRk_le_zero _ hY]
+
+lemma eRk_le_eRk_inter_add_eRk_diff (M : Matroid α) (X Y : Set α) :
+    M.eRk X ≤ M.eRk (X ∩ Y) + M.eRk (X \ Y) := by
+  nth_rw 1 [← inter_union_diff X Y]; apply eRk_union_le_eRk_add_eRk
+
+lemma eRk_le_eRk_add_eRk_diff (M : Matroid α) (h : Y ⊆ X) :
+    M.eRk X ≤ M.eRk Y + M.eRk (X \ Y) := by
+  nth_rw 1 [← union_diff_cancel h]; apply eRk_union_le_eRk_add_eRk
+
+lemma eRk_union_le_encard_add_eRk (M : Matroid α) (X Y : Set α) :
+    M.eRk (X ∪ Y) ≤ X.encard + M.eRk Y :=
+  (M.eRk_union_le_eRk_add_eRk X Y).trans <| add_le_add_right (M.eRk_le_encard _) _
+
+lemma eRk_union_le_eRk_add_encard (M : Matroid α) (X Y : Set α) :
+    M.eRk (X ∪ Y) ≤ M.eRk X + Y.encard :=
+  (M.eRk_union_le_eRk_add_eRk X Y).trans <| add_le_add_left (M.eRk_le_encard _) _
+
+lemma eRank_le_encard_add_eRk_compl (M : Matroid α) (X : Set α) :
+    M.eRank ≤ X.encard + M.eRk (M.E \ X) :=
+  le_trans (by rw [← eRk_inter_ground, eRank_def, union_diff_self,
+    union_inter_cancel_right]) (M.eRk_union_le_encard_add_eRk X (M.E \ X))
+
 end Basic
 
+/-! ### Finiteness -/
+
+@[simp]
+lemma eRank_ne_top_iff (M : Matroid α) : M.eRank ≠ ⊤ ↔ M.RankFinite := by
+  obtain ⟨B, hB⟩ := M.exists_isBase
+  rw [← hB.encard_eq_eRank, encard_ne_top_iff]
+  exact ⟨fun h ↦ hB.rankFinite_of_finite h, fun h ↦ hB.finite⟩
+
+@[deprecated (since := "2025-04-13")] alias rankFinite_iff_eRk_ne_top := eRank_ne_top_iff
+
+lemma eRank_eq_top_iff (M : Matroid α) : M.eRank = ⊤ ↔ M.RankInfinite := by
```

Comment:
I think it could work to have this as simp, provided that `¬ M.RankInfinite` simplifies to `M.RankFinite` (and similarly the other way). And it seems helpful to do so, I think?